### PR TITLE
Extract queryset filtering code from `ReportView` to generic `IndexView`

### PR DESCRIPTION
**Pull request by laymonage**
_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: wagtail/admin/views/mixins.py [hunk 1]
```diff
@@ -10,47 +10,7 @@
 from django.utils.translation import gettext as _
 from xlsxwriter.workbook import Workbook
 
-from wagtail.admin.forms.search import SearchForm
 from wagtail.coreutils import multigetattr
-from wagtail.search.backends import get_search_backend
-from wagtail.search.index import class_is_indexed
-
-
-class SearchableListMixin:
```
Comment (zerolab) [inline wagtail/admin/views/mixins.py:19]: 🤔 this feels like a useful mixin that allows, say, package developers to use without the need `IndexView`. Perhaps we'll want to take stock of the various admin views and interfaces and their use all around?

Note: that this is very unscientific at this point

Comment (laymonage) [reply wagtail/admin/views/mixins.py]: Yeah, I initially didn't remove it but I didn't see it being used anywhere else in Wagtail. I guess we can leave it as it is. I haven't searched whether people use it, though.

File: client/scss/components/_filters.scss [hunk 2]
```diff
@@ -0,0 +1,35 @@
+.has-filters {
```
Comment (laymonage) [inline client/scss/components/_filters.scss:1]: Not a fan of this name, but I can't think of anything else atm.

File: wagtail/contrib/redirects/filters.py [hunk 3]
```diff
@@ -26,3 +27,7 @@ def filter_type(self, queryset, name, value):
         if value and self.request and self.request.user:
             queryset = queryset.filter(is_permanent=value)
         return queryset
+
+    class Meta:
+        model = Redirect
+        exclude = []
```
Comment (laymonage) [inline wagtail/contrib/redirects/filters.py:33]: Without this, the tests fail when calling `self.filters.get_fields()` in the generic `IndexView`: https://github.com/wagtail/wagtail/runs/7384787027?check_suite_focus=true

File: wagtail/admin/views/mixins.py (hunk 1 above)
Comment (kaedroho) [reply wagtail/admin/views/mixins.py]: This code is dead, undocumented, and probably now incompatible with Wagtail's own ``IndexView``. I'm in favour of removing it, the search API should be simple enough to integrate into any custom views.

File: client/src/entrypoints/admin/core.js [hunk 4]
```diff
@@ -318,15 +318,20 @@ $(() => {
         $inputContainer.addClass(workingClasses);
         searchNextIndex++;
         const index = searchNextIndex;
+
+        // Only update q and keep other query params
```
Comment (kaedroho) [inline client/src/entrypoints/admin/core.js:322]: Should we switch back to the first page when the search query changes?

File: wagtail/admin/views/generic/models.py [hunk 5]
```diff
@@ -125,6 +129,63 @@ def get_search_form(self):
...
         )
 
+    def get_queryset(self):
+        self.filters, queryset = self.filter_queryset(super().get_queryset())
+
+        if self.locale:
+            queryset = queryset.filter(locale=self.locale)
+
+        if self.model and issubclass(self.model, DraftStateMixin):
+            queryset.select_related("latest_revision__object_str")
```
Comment (kaedroho) [inline wagtail/admin/views/generic/models.py:139]:
```suggestion
            queryset = queryset.select_related("latest_revision__object_str")
```

**Review by kaedroho — CHANGES_REQUESTED** [review]
A couple of minor tweaks, but otherwise looks good to me!

File: client/src/entrypoints/admin/core.js (hunk 4 above)
Comment (laymonage) [reply client/src/entrypoints/admin/core.js]: Good call. Fixed.

**Review by kaedroho — APPROVED** [review]
All looks good to me now, thanks!

**Pull request merged**